common/utils/src/main/scala/org/apache/spark/internal/Logging.scala

```diff
@@ -182,6 +170,12 @@ trait Logging {
     }
   }
 
+  protected def MDC(key: LogKey, value: Any): MDC = {
+    require(!value.isInstanceOf[MessageWithContext],
+      "the class of value cannot be MessageWithContext")
+    new MDC(key, value)
+  }
```

Comment:
Since it's unable to write a class in Java that allows Scala to construct instances without `new`, I create this method instead.

So, for logging in Scala

```scala
logInfo(log"Application ID is ${MDC(APP_ID, appId)}")
```
previously `MDC(APP_ID, appId)` calls `case class`'s constructor, now it calls this method.